service/init.go

```diff
@@ -85,9 +87,72 @@ func (udr *UDR) Initialize(c *cli.Context) error {
 		return err
 	}
 
+	if os.Getenv("MANAGED_BY_CONFIG_POD") == "true" {
+		logger.InitLog.Infoln("MANAGED_BY_CONFIG_POD is true")
+		go manageGrpcClient(factory.UdrConfig.Configuration.WebuiUri)
+	} else {
+		go func() {
+			logger.InitLog.Infoln("use helm chart config")
+			factory.ConfigPodTrigger <- true
+		}()
+	}
+
 	return nil
 }
 
+// manageGrpcClient connects the config pod GRPC server and subscribes the config changes.
+// Then it updates UDR configuration.
+func manageGrpcClient(webuiUri string) {
+	var configChannel chan *protos.NetworkSliceResponse
+	var client grpcClient.ConfClient
+	var stream protos.ConfigService_NetworkSliceSubscribeClient
+	var err error
+	count := 0
+	for {
+		if client != nil {
+			if client.CheckGrpcConnectivity() != "ready" {
+				time.Sleep(time.Second * 30)
+				count++
+				if count > 5 {
+					err = client.GetConfigClientConn().Close()
+					if err != nil {
+						logger.InitLog.Infof("failing ConfigClient is not closed properly: %+v", err)
+					}
+					client = nil
+					count = 0
+				}
+				logger.InitLog.Infoln("checking the connectivity readiness")
+				continue
+			}
+
+			if stream == nil {
+				stream, err = client.SubscribeToConfigServer()
+				if err != nil {
+					logger.InitLog.Infof("failing SubscribeToConfigServer: %+v", err)
+					continue
+				}
+			}
+
+			if configChannel == nil {
+				configChannel = client.PublishOnConfigChange(true, stream)
+				logger.InitLog.Infoln("PublishOnConfigChange is triggered")
+				factory.ConfigUpdateDbTrigger = make(chan *factory.UpdateDb, 10)
```

Comment:
@gatici, @thakurajayL,
I found the reason why this PR is not working for me, and it is because of this line `factory.ConfigUpdateDbTrigger = make(chan *factory.UpdateDb, 10)`. When I use this PR as-is, I get a kind of GRPC connection issue as you can see below:
```
...
...
2024-11-05T17:56:39.963Z	INFO	factory/config.go:118	site name enterprise	{"component": "UDR", "category": "GRPC"}
2024-11-05T17:56:39.963Z	INFO	factory/config.go:120	plmn mcc 208	{"component": "UDR", "category": "GRPC"}
2024/11/05 17:56:40 INFO: [core] [Channel #1]Resolver state updated: {
  "Addresses": [
    {
      "Addr": "10.42.205.9:9876",
      "ServerName": "",
      "Attributes": null,
      "BalancerAttributes": null,
      "Metadata": null
    }
  ],
  "Endpoints": [
    {
      "Addresses": [
        {
          "Addr": "10.42.205.9:9876",
          "ServerName": "",
          "Attributes": null,
          "BalancerAttributes": null,
          "Metadata": null
        }
      ],
      "Attributes": null
    }
  ],
  "ServiceConfig": null,
  "Attributes": null
} ()
2024/11/05 17:56:40 INFO: [core] [Channel #1 SubChannel #2]addrConn: updateAddrs addrs (1 of 1): [{Addr: "10.42.205.9:9876", ServerName: "", }]
2024/11/05 17:57:10 INFO: [core] [Channel #1]Resolver state updated: {
  "Addresses": [
    {
      "Addr": "10.42.205.9:9876",
      "ServerName": "",
      "Attributes": null,
      "BalancerAttributes": null,
      "Metadata": null
    }
  ],
  "Endpoints": [
    {
      "Addresses": [
        {
          "Addr": "10.42.205.9:9876",
          "ServerName": "",
          "Attributes": null,
          "BalancerAttributes": null,
          "Metadata": null
        }
      ],
      "Attributes": null
    }
  ],
  "ServiceConfig": null,
  "Attributes": null
} ()
2024/11/05 17:57:10 INFO: [core] [Channel #1 SubChannel #2]addrConn: updateAddrs addrs (1 of 1): [{Addr: "10.42.205.9:9876", ServerName: "", }]
```

Before this PR, the UDR log shows the following (`config update DB trigger`) right after the `plmn mcc 208` log message
```
...
...
2024-11-05T18:34:29.090Z	INFO	factory/config.go:118	site name enterprise	{"component": "UDR", "category": "GRPC"}
2024-11-05T18:34:29.090Z	INFO	factory/config.go:120	plmn mcc 208	{"component": "UDR", "category": "GRPC"}
2024-11-05T18:34:29.090Z	INFO	service/init.go:267	config update DB trigger	{"component": "UDR", "category": "Init"}
...
...
```

When digging deeper into this, I see that `configUpdateDb()` gets called before the `ConfigUpdateDbTrigger` has an assigned value. And that is the reason that the E2E tests fail for me. Below are two options to address this issue (my personal preference is the first option):

```diff
diff --git a/factory/config.go b/factory/config.go
index 669e222..9b4c3e5 100644
--- a/factory/config.go
+++ b/factory/config.go
@@ -80,6 +80,7 @@ var (

 func init() {
        ConfigPodTrigger = make(chan bool)
+       ConfigUpdateDbTrigger = make(chan *UpdateDb, 10)
 }

 func (c *Config) GetVersion() string {
diff --git a/service/init.go b/service/init.go
index 4edebe7..146abc2 100644
--- a/service/init.go
+++ b/service/init.go
@@ -136,7 +136,6 @@ func manageGrpcClient(webuiUri string) {
                        if configChannel == nil {
                                configChannel = client.PublishOnConfigChange(true, stream)
                                logger.InitLog.Infoln("PublishOnConfigChange is triggered")
-                               factory.ConfigUpdateDbTrigger = make(chan *factory.UpdateDb, 10)
                                go factory.UdrConfig.UpdateConfig(configChannel, factory.ConfigUpdateDbTrigger)
                                logger.InitLog.Infoln("UDR updateConfig is triggered")
                        }
```
or
```diff
diff --git a/service/init.go b/service/init.go
index 4edebe7..809b4ee 100644
--- a/service/init.go
+++ b/service/init.go
@@ -89,6 +89,7 @@ func (udr *UDR) Initialize(c *cli.Context) error {

        if os.Getenv("MANAGED_BY_CONFIG_POD") == "true" {
                logger.InitLog.Infoln("MANAGED_BY_CONFIG_POD is true")
+               factory.ConfigUpdateDbTrigger = make(chan *factory.UpdateDb, 10)
                go manageGrpcClient(factory.UdrConfig.Configuration.WebuiUri)
        } else {
                go func() {
@@ -136,7 +137,6 @@ func manageGrpcClient(webuiUri string) {
                        if configChannel == nil {
                                configChannel = client.PublishOnConfigChange(true, stream)
                                logger.InitLog.Infoln("PublishOnConfigChange is triggered")
-                               factory.ConfigUpdateDbTrigger = make(chan *factory.UpdateDb, 10)
                                go factory.UdrConfig.UpdateConfig(configChannel, factory.ConfigUpdateDbTrigger)
                                logger.InitLog.Infoln("UDR updateConfig is triggered")
                        }
```

After applying one of the above proposals on top of your PR, the E2E tests successfully complete as you can see below:
```
        "Profile Name: profile1 , Profile Type: register",
        "Ue's Passed: 5 , Ue's Failed: 0",
        "Profile Status: PASS",
        "Profile Name: profile7 , Profile Type: uereqpdusessrelease",
        "Ue's Passed: 5 , Ue's Failed: 0",
        "Profile Status: PASS",
        "Profile Name: profile2 , Profile Type: pdusessest",
        "Ue's Passed: 5 , Ue's Failed: 0",
        "Profile Status: PASS",
        "Profile Name: profile5 , Profile Type: deregister",
        "Ue's Passed: 5 , Ue's Failed: 0",
        "Profile Status: PASS",
        "Profile Name: profile3 , Profile Type: anrelease",
        "Ue's Passed: 5 , Ue's Failed: 0",
        "Profile Status: PASS",
        "Profile Name: profile4 , Profile Type: uetriggservicereq",
        "Ue's Passed: 5 , Ue's Failed: 0",
        "Profile Status: PASS"
```

Reply:
Hello Gabriel, 
It applied the option 1. 
Many thanks!